File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.java
import java.util.*;

class Solution {
    public int[] solution(int[][] edges) {
        Map<Integer, List<Integer>> graphOut = new HashMap<>();
        Map<Integer, List<Integer>> graphIn = new HashMap<>();
        Set<Integer> points = new HashSet<>();

        for (int[] edge : edges) {
            int a = edge[0];
            int b = edge[1];
            graphOut.computeIfAbsent(a, k -> new ArrayList<>()).add(b);
            graphIn.computeIfAbsent(b, k -> new ArrayList<>()).add(a);
            points.add(a);
            points.add(b);
        }

        int created = 0;
        for (int point : points) {
            if (!graphIn.containsKey(point) && graphOut.getOrDefault(point, new ArrayList<>()).size() >= 2) {
                if (created == 0 || graphOut.get(created).size() < graphOut.get(point).size()) {
                    created = point;
                }
            }
        }

        int[] result = new int[4];
        result[0] = created;

        for (int start : graphOut.get(created)) {
            int next = start;
            Set<Integer> visited = new HashSet<>();
            while (true) {
                if (graphOut.getOrDefault(next, new ArrayList<>()).size() == 2) {
                    result[3]++;
                    break;
                } else if (graphOut.getOrDefault(next, new ArrayList<>()).size() == 1) {
                    if (visited.contains(next)) {
                        result[1]++;
                        break;
                    }
                    visited.add(next);
                    next = graphOut.get(next).get(0);
                } else {
                    result[2]++;
                    break;
                }
            }
        }

        return result;
    }
}

File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.py
from collections import defaultdict

def solution(edges):
    graph_in = defaultdict(list)
    graph_out = defaultdict(list)
    points = set()
    for a,b in edges:
        graph_out[a].append(b)
        graph_in[b].append(a)
        points.add(a)
        points.add(b)

    
    created = 0
    for dot_number in points:
        if dot_number not in graph_out and dot_number not in graph_in:
            break
        if len(graph_out[dot_number])>=2 and dot_number not in graph_in:
            if created==0 or len(graph_out[created])<len(graph_out[dot_number]):
                created = dot_number
                break
    
    
    result = [created,0,0,0]            
    for start in graph_out[created]:
        next = start
        while True:
            if len(graph_out[next]) == 2:
                result[3]+=1
                break
    
            elif len(graph_out[next]) == 1:
                next=graph_out[next][0]
                if next == start: # 도넛
                    result[1]+=1
                    break
            elif not graph_out[next]: # 막대
                result[2]+=1
                break
                
    return result